src/include/Context.h

```diff
@@ -56,20 +56,34 @@ using GenContextURef = std::unique_ptr<GenContext<T> >;
 /*
  * Context - abstract callback class
  */
+class Finisher;
 class Context {
   Context(const Context& other);
   const Context& operator=(const Context& other);
 
  protected:
   virtual void finish(int r) = 0;
 
+  // variant of finish that is safe to call "synchronously."  override should
+  // return true.
```

Comment:
There ought to be a way to build this interface so that it doesn't require returning the correct value. (One terrible way: a private member we set in the default implementation indicating "not implemented" and check for when invoking it.)

Also, it needs more documentation in the source. What other constraints exist?